src/vm/diagnosticprotocolhelper.cpp

```diff
@@ -88,4 +93,137 @@ void DiagnosticProtocolHelper::GenerateCoreDump(IpcStream* pStream)
+
+    uint8_t *pBufferCursor = 0;
+    uint32_t bufferLen = 0;
+    uint32_t pathSize = 0;
+    uint32_t bytesToRead = 0;
+    uint32_t bufferBytesRead = 0;
+    uint8_t *pClientDataCursor = 0;
+    
+    bool fSuccess = pStream->Read(buffer, sizeof(buffer), nNumberOfBytesRead);
+    if (!fSuccess)
+    {
+        hr = E_UNEXPECTED;
+        goto ErrExit;
+    }
+
+    pBufferCursor = buffer;
+    bufferLen = nNumberOfBytesRead;
+    if (!(TryParse(pBufferCursor, bufferLen, dwAttachTimeout) &&
+        TryParse(pBufferCursor, bufferLen, profilerGuid)))
+    {
+        hr = E_INVALIDARG;
+        goto ErrExit;
+    }
+
+    if (!TryParse(pBufferCursor, bufferLen, cbProfilerPath) ||
+        cbProfilerPath > bufferLen)
+    {
+        // TODO: A really long path (thousands of characters) could be longer than 
+        // the buffer here, if that happens we'll return E_INVALIDARG. The read
+        // buffer is 8192 bytes, so has room for 4096 16 bit characters. Minus a few for
+        // the header, etc, the realistic max is around 4000 characters.
+        hr = E_INVALIDARG;
+        goto ErrExit;
+    }
+
+    pwszProfilerPath = new (nothrow) WCHAR[cbProfilerPath];
+    if (pwszProfilerPath == nullptr)
+    {
+        hr = E_INVALIDARG;
+        goto ErrExit;
+    }
+
+    pathSize = cbProfilerPath * sizeof(WCHAR);
+    memcpy(pwszProfilerPath, pBufferCursor, pathSize);
+    bufferLen -= pathSize;
+    pBufferCursor += pathSize;
+        
+    if (!TryParse(pBufferCursor, bufferLen, cbClientData))
+    {
+        hr = E_INVALIDARG;
+        goto ErrExit;
+    }
+
+    pClientData = new (nothrow) uint8_t[cbClientData];
+    if (pClientData == nullptr)
+    {
+        hr = E_OUTOFMEMORY;
+        goto ErrExit;
+    }
+
+    bufferBytesRead = 0;
+    pClientDataCursor = pClientData;
+    // TODO: get rid of this ad-hoc byte[] parsing code
+    while (bufferBytesRead < cbClientData)
+    {
+        if (bufferLen == 0)
+        {
+            // Client data was bigger than the buffer, need to read more
+            fSuccess = pStream->Read(buffer, sizeof(buffer), nNumberOfBytesRead);
+            if (!fSuccess)
+            {
+                hr = E_UNEXPECTED;
+                goto ErrExit;
+            }
+
+            pBufferCursor = buffer;
+            bufferLen = nNumberOfBytesRead;
+        }
+
+        bytesToRead = min((cbClientData - bufferBytesRead), bufferLen);
+        memcpy(pClientDataCursor, pBufferCursor, bytesToRead);
+        pClientDataCursor += bytesToRead;
+        
+        _ASSERTE(bytesToRead <= bufferLen && "bytesToRead > bufferLen means we overran the buffer");
+        bufferLen -= bytesToRead;
+        bufferBytesRead += bytesToRead;
+    }
+
+    _ASSERTE(bufferBytesRead == cbClientData && "bufferBytesRead > cbClientData means we read too far");
+
+    if (cbClientData == 0)
+    {
+        pClientData = nullptr;
+    }
+
+    hr = ProfilingAPIUtility::LoadProfilerForAttach(&profilerGuid,
```

Comment:
As a second issue, its safe for now but @josalem will need to coordinate with you if he makes the diagnostic server multi-threaded. LoadProfilerForAttach doesn't appear safe to run in a multi-threaded environment without adding some additional synchronization that would let an attach request cleanly win/lose the attach race.

Reply:
Since this is the only place we would be loading a profiler from, I think all that would be necessary is to add a Crst around the call.

Now that I'm thinking about synchronization, it may race with a startup profiler too. I'll have to look in to it more tomorrow.